kong/runloop/plugin_servers/plugin.lua

```diff
@@ -0,0 +1,346 @@
+
+else
+  local NOOP = function() end
+
+  req_start_time   = NOOP
+  req_get_headers  = NOOP
+  resp_get_headers = NOOP
+end
+
+local function get_saved()
+  return save_for_later[coroutine_running()]
+end
+
+local exposed_api = {
+  kong = kong,
+
+  get_saved_for_later = get_saved,
+
+  ["kong.log.serialize"] = function()
+    local saved = get_saved()
+    return cjson_encode(saved and saved.serialize_data or kong.log.serialize())
+  end,
+
+  ["kong.nginx.get_var"] = function(v)
+    return ngx_var[v]
+  end,
+
+  ["kong.nginx.get_tls1_version_str"] = ngx_ssl.get_tls1_version_str,
+
+  ["kong.nginx.get_ctx"] = function(k)
+    local saved = get_saved()
+    local ngx_ctx = saved and saved.ngx_ctx or ngx.ctx
+    return ngx_ctx[k]
+  end,
+
+  ["kong.nginx.set_ctx"] = function(k, v)
+    local saved = get_saved()
+    local ngx_ctx = saved and saved.ngx_ctx or ngx.ctx
+    ngx_ctx[k] = v
+  end,
+
+  ["kong.ctx.shared.get"] = function(k)
+    local saved = get_saved()
+    local ctx_shared = saved and saved.ctx_shared or kong.ctx.shared
+    return ctx_shared[k]
+  end,
+
+  ["kong.ctx.shared.set"] = function(k, v)
+    local saved = get_saved()
+    local ctx_shared = saved and saved.ctx_shared or kong.ctx.shared
+    ctx_shared[k] = v
+  end,
+
+  ["kong.request.get_headers"] = function(max)
+    local saved = get_saved()
+    return saved and saved.request_headers or kong.request.get_headers(max)
+  end,
+
+  ["kong.request.get_header"] = function(name)
+    local saved = get_saved()
+    if not saved then
+      return kong.request.get_header(name)
+    end
+
+    local header_value = saved.request_headers[name]
+    if type(header_value) == "table" then
+      header_value = header_value[1]
+    end
+
+    return header_value
+  end,
+
+  ["kong.request.get_uri_captures"] = function()
+    local saved = get_saved()
+    local ngx_ctx = saved and saved.ngx_ctx or ngx.ctx
+    return kong.request.get_uri_captures(ngx_ctx)
+  end,
+
+  ["kong.response.get_status"] = function()
+    local saved = get_saved()
+    return saved and saved.response_status or kong.response.get_status()
+  end,
+
+  ["kong.response.get_headers"] = function(max)
+    local saved = get_saved()
+    return saved and saved.response_headers or kong.response.get_headers(max)
+  end,
+
+  ["kong.response.get_header"] = function(name)
+    local saved = get_saved()
+    if not saved then
+      return kong.response.get_header(name)
+    end
+
+    local header_value = saved.response_headers and saved.response_headers[name]
+    if type(header_value) == "table" then
+      header_value = header_value[1]
+    end
+
+    return header_value
+  end,
+
+  ["kong.response.get_source"] = function()
+    local saved = get_saved()
+    return kong.response.get_source(saved and saved.ngx_ctx or nil)
+  end,
+
+  ["kong.nginx.req_start_time"] = function()
+    local saved = get_saved()
+    return saved and saved.req_start_time or req_start_time()
+  end,
+}
```

Comment:
It seems like almost all of the handler functions require access to the saved request data.  Would it make sense to pass it to the handler functions as `request` parameter rather than having them all call `get_saved()`?  I'm also not thrilled about `saved` and `get_saved` as names.  Would `req` and `get_req` not be more appropriate?

Reply:
It's more like "retrieve_request_context" here.